README.md
<!-- [![pub package](https://img.shields.io/pub/v/in_app_update.svg)](https://pub.dev/packages/in_app_update) -->

Presented by

<!-- <img src="https://ffuf.de/assets/img/shared/ffuf-logo-red.svg" width="150"> -->

https://lapakprogrammer.com/

Maintained by [ACHMAD MAULANA](https://lapakprogrammer.com/developer/achmad-maulana)

# in_googleplay_update

Enables In App Updates on Android using the official Android APIs.

https://developer.android.com/guide/app-bundle/in-app-updates

<img src="https://2.bp.blogspot.com/-9V4ZsdRRnIA/XNSYN-do_OI/AAAAAAAAI90/2yFBsTij0kcibkGRuB79fS_jZKcy-APdQCLcBGAs/s1600/Screen%2BShot%2B2019-05-09%2Bat%2B2.13.58%2BPM.png" width="400">

## Documentation

The following methods are exposed:
- `Future<AppUpdateInfo> checkForUpdate()`: Checks if there's an update available
- `Future<void> performImmediateUpdate()`: Performs an immediate update (full-screen)
- `Future<void> startFlexibleUpdate()`: Starts a flexible update (background download)
- `Future<void> completeFlexibleUpdate()`: Actually installs an available flexible update
- `Stream<String> flexibleUpdateStream()`: To listener flexible update state this return json string `bytes_downloaded: string, total_bytes_to_download` used this for make progress har download 
example: flexible progressbar
```dart
StreamBuilder<String>(
  stream: InAppPlayUpdate.flexibleUpdateStream(),
  builder: (context, snapshot) {
    if (snapshot.hasData) {
      String byteStringDownload = snapshot.data ?? '{}';
      final byteDownload = json.decode(byteStringDownload);
      return Column(
        children: [
          if(int.parse(byteDownload['bytes_downloaded']) == 0 && int.parse(byteDownload['total_bytes_to_download']) == 0)...[
            ElevatedButton(
              onPressed: () {
                InAppPlayUpdate.completeFlexibleUpdate().then((_) {
                  showSnack("Success!");
                }).catchError((e) {
                  showSnack(e.toString());
                });
              },
              child: const Text('Complete flexible update'),
            ),
          ] else...[
            LinearProgressIndicator(
              value: (int.parse(byteDownload['bytes_downloaded']) / int.parse(byteDownload['total_bytes_to_download'])),
            ),
            Text('${filesize(int.parse(byteDownload['bytes_downloaded']))}/${filesize(int.parse(byteDownload['total_bytes_to_download']))}',
              style: const TextStyle(
                  fontSize: 14
              ),
            )
          ]
        ],
      );
    } else {
      return Column(
        children: [
          SizedBox(height: 10),
          SizedBox(
            width: 30,
            height: 30,
            child: CircularProgressIndicator(),
          ),
          Text('This loading for download flexible indicator',
            style: const TextStyle(
                fontSize: 14
            ),
          )
        ],
      );
    }
  },
)
```

Please have a look in the example app on how to use it!

### Android

This plugin integrates the official Android APIs to perform in app updated that were released in 2019:
https://developer.android.com/guide/app-bundle/in-app-updates

# Troubleshooting

## Getting ERROR_API_NOT_AVAILABLE error
Be aware that this plugin cannot be tested locally. It must be installed via Google Play to work. 
Please check the official documentation about In App Updates from Google.

CHANGELOG.md
## 0.0.1

* TODO: release a version 0.0.1

## 0.0.2

* TODO: update minor fix some issue from version 0.0.1

## 0.0.3

* TODO: update minor fix some issue from version 0.0.2

## 0.0.4

* TODO: update minor fix some issue from version 0.0.3


## 0.0.5

* TODO: update minor fix some issue from version 0.0.4

## 0.0.6

* TODO: update minor fix some issue from version 0.0.5

## 0.0.7

* TODO: update minor fix some issue from version 0.0.6


## 0.0.8

* TODO: update minor fix some issue from version 0.0.7

## 0.0.9

* TODO: update stream listener for get progress byte download for flexible update
- bytes_downloaded: string  [number far of number byte of download]
- total_bytes_to_download: string 

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:in_googleplay_update_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/in_googleplay_update_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:in_googleplay_update/in_googleplay_update.dart';
import 'package:in_googleplay_update/in_googleplay_update_platform_interface.dart';
import 'package:in_googleplay_update/in_googleplay_update_method_channel.dart';
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

class MockInGoogleplayUpdatePlatform
    with MockPlatformInterfaceMixin
    implements InGoogleplayUpdatePlatform {

  @override
  Future<String?> getPlatformVersion() => Future.value('42');
}

void main() {
  final InGoogleplayUpdatePlatform initialPlatform = InGoogleplayUpdatePlatform.instance;

  test('$MethodChannelInGoogleplayUpdate is the default instance', () {
    expect(initialPlatform, isInstanceOf<MethodChannelInGoogleplayUpdate>());
  });

  test('getPlatformVersion', () async {
    InAppPlayUpdate inGoogleplayUpdatePlugin = InAppPlayUpdate();
    MockInGoogleplayUpdatePlatform fakePlatform = MockInGoogleplayUpdatePlatform();
    InGoogleplayUpdatePlatform.instance = fakePlatform;

    expect(await InAppPlayUpdate.startFlexibleUpdate(), '42');
  });
}


test/in_googleplay_update_method_channel_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:in_googleplay_update/in_googleplay_update_method_channel.dart';

void main() {
  MethodChannelInGoogleplayUpdate platform = MethodChannelInGoogleplayUpdate();
  const MethodChannel channel = MethodChannel('in_googleplay_update');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await platform.getPlatformVersion(), '42');
  });
}


lib/in_googleplay_update.dart
import 'dart:async';

import 'package:flutter/services.dart';

class InstallStatus {
  InstallStatus._();

  static int get unknown => 0;
  static int get pending => 1;
  static int get downloading => 2;
  static int get installing => 3;
  static int get installed => 4;
  static int get failed => 5;
  static int get canceled => 6;
  static int get downloaded => 11;
}

class UpdateAvailability {
  UpdateAvailability._();

  static int get unknown => 0;
  static int get updateNotAvailable => 1;
  static int get updateAvailable => 2;
  static int get developerTriggeredUpdateInProgress => 3;
}

enum AppUpdateResult {
  /// The user has accepted the update. For immediate updates, you might not receive this callback because the update should already be completed by Google Play by the time the control is given back to your app.
  success,

  /// The user has denied or cancelled the update.
  userDeniedUpdate,

  /// Some other error prevented either the user from providing consent or the update to proceed.
  inAppUpdateFailed,
}

class InAppPlayUpdate {
  static const MethodChannel _channel = MethodChannel('in_googleplay_update');
  static const EventChannel _listener =
      EventChannel('in_googleplay_update_event');

  static Stream<String> flexibleUpdateStream() {
    return _listener.receiveBroadcastStream().map((event) => event.toString());
  }

  /// Returns [AppUpdateInfo], which can be used to decide if
  /// [startFlexibleUpdate] or [performImmediateUpdate] should be called.
  static Future<AppPlayUpdateInfo> checkForUpdate() async {
    final result = await _channel.invokeMethod('checkForUpdate');

    return AppPlayUpdateInfo(
      result['updateAvailability'],
      result['immediateAllowed'],
      result['flexibleAllowed'],
      result['availableVersionCode'],
      result['installStatus'],
      result['packageName'],
      result['clientVersionStalenessDays'],
      result['updatePriority'],
    );
  }

  /// Performs an immediate update that is entirely handled by the Play API.
  ///
  /// [checkForUpdate] has to be called first to be able to run this.
  static Future<AppUpdateResult> performImmediateUpdate() async {
    try {
      await _channel.invokeMethod('performImmediateUpdate');
      return AppUpdateResult.success;
    } on PlatformException catch (e) {
      if (e.code == 'USER_DENIED_UPDATE') {
        return AppUpdateResult.userDeniedUpdate;
      } else if (e.code == 'IN_APP_UPDATE_FAILED') {
        return AppUpdateResult.inAppUpdateFailed;
      }

      throw e;
    }
  }

  /// Starts the download of the app update.
  ///
  /// Throws a [PlatformException] if the download fails.
  /// When the returned [Future] is completed without any errors,
  /// [completeFlexibleUpdate] can be called to install the update.
  ///
  /// [checkForUpdate] has to be called first to be able to run this.
  static Future<AppUpdateResult> startFlexibleUpdate() async {
    try {
      await _channel.invokeMethod('startFlexibleUpdate');
      return AppUpdateResult.success;
    } on PlatformException catch (e) {
      if (e.code == 'USER_DENIED_UPDATE') {
        return AppUpdateResult.userDeniedUpdate;
      } else if (e.code == 'IN_APP_UPDATE_FAILED') {
        return AppUpdateResult.inAppUpdateFailed;
      }

      throw e;
    }
  }

  /// Installs the update downloaded via [startFlexibleUpdate].
  /// [startFlexibleUpdate] has to be completed successfully.
  static Future<void> completeFlexibleUpdate() async {
    return await _channel.invokeMethod('completeFlexibleUpdate');
  }

  /// Installs the update status progress via [startFlexibleUpdate].
  /// [checkFlexibleUpdateProgress] has to be check progress value.
  static Future<AppUpdateResult> checkFlexibleUpdateProgress() async {
    try {
      await _channel.invokeMethod('checkFlexibleUpdateProgress');
      return AppUpdateResult.success;
    } on PlatformException catch (e) {
      if (e.code == 'USER_DENIED_UPDATE') {
        return AppUpdateResult.userDeniedUpdate;
      } else if (e.code == 'IN_APP_UPDATE_FAILED') {
        return AppUpdateResult.inAppUpdateFailed;
      }

      throw e;
    }
  }
}

class AppUpdateListener {
  final String? byteDownload;
  final String? totalByteDownload;

  AppUpdateListener({
    this.byteDownload,
    this.totalByteDownload,
  });
}

class AppPlayUpdateInfo {
  final int updateAvailability;
  final bool immediateUpdateAllowed, flexibleUpdateAllowed;
  final int? availableVersionCode;
  final int installStatus;
  final String packageName;
  final int updatePriority;
  final int? clientVersionStalenessDays;

  AppPlayUpdateInfo(
    this.updateAvailability,
    this.immediateUpdateAllowed,
    this.flexibleUpdateAllowed,
    this.availableVersionCode,
    this.installStatus,
    this.packageName,
    this.clientVersionStalenessDays,
    this.updatePriority,
  );

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is AppPlayUpdateInfo &&
          runtimeType == other.runtimeType &&
          updateAvailability == other.updateAvailability &&
          immediateUpdateAllowed == other.immediateUpdateAllowed &&
          flexibleUpdateAllowed == other.flexibleUpdateAllowed &&
          availableVersionCode == other.availableVersionCode &&
          installStatus == other.installStatus &&
          packageName == other.packageName &&
          clientVersionStalenessDays == other.clientVersionStalenessDays &&
          updatePriority == other.updatePriority;

  @override
  int get hashCode =>
      updateAvailability.hashCode ^
      immediateUpdateAllowed.hashCode ^
      flexibleUpdateAllowed.hashCode ^
      availableVersionCode.hashCode ^
      installStatus.hashCode ^
      packageName.hashCode ^
      clientVersionStalenessDays.hashCode ^
      updatePriority.hashCode;

  @override
  String toString() =>
      'InAppUpdateState{updateAvailability: $updateAvailability, '
      'immediateUpdateAllowed: $immediateUpdateAllowed, '
      'flexibleUpdateAllowed: $flexibleUpdateAllowed, '
      'availableVersionCode: $availableVersionCode, '
      'installStatus: $installStatus, '
      'packageName: $packageName, '
      'clientVersionStalenessDays: $clientVersionStalenessDays, '
      'updatePriority: $updatePriority}';
}


lib/in_googleplay_update_platform_interface.dart
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

import 'in_googleplay_update_method_channel.dart';

abstract class InGoogleplayUpdatePlatform extends PlatformInterface {
  /// Constructs a InGoogleplayUpdatePlatform.
  InGoogleplayUpdatePlatform() : super(token: _token);

  static final Object _token = Object();

  static InGoogleplayUpdatePlatform _instance = MethodChannelInGoogleplayUpdate();

  /// The default instance of [InGoogleplayUpdatePlatform] to use.
  ///
  /// Defaults to [MethodChannelInGoogleplayUpdate].
  static InGoogleplayUpdatePlatform get instance => _instance;

  /// Platform-specific implementations should set this with their own
  /// platform-specific class that extends [InGoogleplayUpdatePlatform] when
  /// they register themselves.
  static set instance(InGoogleplayUpdatePlatform instance) {
    PlatformInterface.verifyToken(instance, _token);
    _instance = instance;
  }

  Future<String?> getPlatformVersion() {
    throw UnimplementedError('platformVersion() has not been implemented.');
  }
}


android/src/main/kotlin/com/lapakprogrammer/ingoogleplayupdate/in_googleplay_update/InGoogleplayUpdatePlugin.kt
package com.lapakprogrammer.ingoogleplayupdate.in_googleplay_update

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Activity.RESULT_CANCELED
import android.app.Activity.RESULT_OK
import android.app.Application
import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.annotation.NonNull
import com.google.android.play.core.appupdate.AppUpdateInfo
import com.google.android.play.core.appupdate.AppUpdateManager
import com.google.android.play.core.appupdate.AppUpdateManagerFactory
import com.google.android.play.core.install.InstallStateUpdatedListener
import com.google.android.play.core.install.model.*
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.*
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.google.gson.Gson

interface ActivityProvider {
  fun addActivityResultListener(callback: PluginRegistry.ActivityResultListener)
  fun activity(): Activity?
}

class InGoogleplayUpdatePlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler,
  PluginRegistry.ActivityResultListener, Application.ActivityLifecycleCallbacks, ActivityAware {

  companion object {
    private const val REQUEST_CODE_START_UPDATE = 1276
  }

  private lateinit var channel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private var appUpdateEventSink: EventChannel.EventSink? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(
      flutterPluginBinding.binaryMessenger,
      "in_googleplay_update"
    )
    channel.setMethodCallHandler(this)

    eventChannel = EventChannel(
      flutterPluginBinding.binaryMessenger,
      "in_googleplay_update_event"
    )
    eventChannel.setStreamHandler(this)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  private var activityProvider: ActivityProvider? = null
  private var updateResult: Result? = null
  private var appUpdateType: Int? = null
  private var appUpdateInfo: AppUpdateInfo? = null
  private var appUpdateManager: AppUpdateManager? = null

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "checkForUpdate" -> checkForUpdate(result)
      "performImmediateUpdate" -> performImmediateUpdate(result)
      "startFlexibleUpdate" -> startFlexibleUpdate(result)
      "completeFlexibleUpdate" -> completeFlexibleUpdate(result)
      else -> result.notImplemented()
    }
  }

  override fun onCancel(arguments: Any?) {
    appUpdateEventSink = null
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    appUpdateEventSink = events
  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
    if (requestCode == REQUEST_CODE_START_UPDATE) {
      if (appUpdateType == AppUpdateType.IMMEDIATE) {
        if (resultCode == RESULT_CANCELED) {
          updateResult?.error("USER_DENIED_UPDATE", resultCode.toString(), null)
        } else if (resultCode == RESULT_OK) {
          updateResult?.success(null)
        } else if (resultCode == ActivityResult.RESULT_IN_APP_UPDATE_FAILED) {
          updateResult?.error("IN_APP_UPDATE_FAILED", "Some other error prevented either the user from providing consent or the update to proceed.", null)
        }
        updateResult = null
        return true
      }else if (appUpdateType == AppUpdateType.FLEXIBLE) {
        if (resultCode == RESULT_CANCELED) {
          updateResult?.error("USER_DENIED_UPDATE", resultCode.toString(), null)
          updateResult = null
        }
        else if (resultCode == ActivityResult.RESULT_IN_APP_UPDATE_FAILED) {
          updateResult?.error("IN_APP_UPDATE_FAILED", resultCode.toString(), null)
          updateResult = null
        }
        return true
      }
    }
    return false
  }

  override fun onAttachedToActivity(activityPluginBinding: ActivityPluginBinding) {
    activityProvider = object : ActivityProvider {
      override fun addActivityResultListener(callback: PluginRegistry.ActivityResultListener) {
        activityPluginBinding.addActivityResultListener(callback)
      }

      override fun activity(): Activity? {
        return activityPluginBinding.activity
      }
    }
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activityProvider = null
  }

  override fun onReattachedToActivityForConfigChanges(activityPluginBinding: ActivityPluginBinding) {
    activityProvider = object : ActivityProvider {
      override fun addActivityResultListener(callback: PluginRegistry.ActivityResultListener) {
        activityPluginBinding.addActivityResultListener(callback)
      }

      override fun activity(): Activity? {
        return activityPluginBinding.activity
      }
    }
  }

  override fun onDetachedFromActivity() {
    activityProvider = null
  }

  override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}

  override fun onActivityPaused(activity: Activity) {}

  override fun onActivityStarted(activity: Activity) {}

  override fun onActivityDestroyed(activity: Activity) {}

  override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}

  override fun onActivityStopped(activity: Activity) {}

  override fun onActivityResumed(activity: Activity) {
    appUpdateManager
      ?.appUpdateInfo
      ?.addOnSuccessListener { appUpdateInfo ->
        if (appUpdateInfo.updateAvailability()
          == UpdateAvailability.DEVELOPER_TRIGGERED_UPDATE_IN_PROGRESS
          && appUpdateType == AppUpdateType.IMMEDIATE
        ) {
          appUpdateManager?.startUpdateFlowForResult(
            appUpdateInfo,
            AppUpdateType.IMMEDIATE,
            activity,
            REQUEST_CODE_START_UPDATE
          )
        }
      }
  }

  private fun performImmediateUpdate(result: Result) = checkAppState(result) {
    appUpdateType = AppUpdateType.IMMEDIATE
    updateResult = result
    appUpdateManager?.startUpdateFlowForResult(
      appUpdateInfo!!,
      AppUpdateType.IMMEDIATE,
      activityProvider!!.activity()!!,
      REQUEST_CODE_START_UPDATE
    )
  }

  private fun checkAppState(result: Result, block: () -> Unit) {
    requireNotNull(appUpdateInfo) {
      result.error("REQUIRE_CHECK_FOR_UPDATE", "Call checkForUpdate first!", null)
    }
    requireNotNull(activityProvider?.activity()) {
      result.error("REQUIRE_FOREGROUND_ACTIVITY", "in_googleplay_update requires a foreground activity", null)
    }
    requireNotNull(appUpdateManager) {
      result.error("REQUIRE_CHECK_FOR_UPDATE", "Call checkForUpdate first!", null)
    }
    block()
  }

  @SuppressLint("LongLogTag")
  private fun startFlexibleUpdate(result: Result) = checkAppState(result) {
    appUpdateType = AppUpdateType.FLEXIBLE
    updateResult = result
    appUpdateManager?.startUpdateFlowForResult(
      appUpdateInfo!!,
      AppUpdateType.FLEXIBLE,
      activityProvider!!.activity()!!,
      REQUEST_CODE_START_UPDATE
    )

    val installStateUpdatedListener = InstallStateUpdatedListener { state ->
      // (Optional) Provide a download progress bar.
      if (state.installStatus() == InstallStatus.DOWNLOADING) {

        Log.d("FLEXIBLE_PROGRESS_UPDATE byte:", state.bytesDownloaded().toString());
        Log.d("FLEXIBLE_PROGRESS_UPDATE total byte:", state.totalBytesToDownload().toString());

        val byteData = InGooglePlayByteData(
          state.bytesDownloaded().toString(),
          state.totalBytesToDownload().toString()
        );

        val byteJsonString = Gson().toJson(byteData)

        appUpdateEventSink?.success(byteJsonString);

        updateResult?.success(null)
        updateResult = null
        // Show update progress bar.
      } else  if (state.installStatus() == InstallStatus.DOWNLOADED) {


        Log.d("FLEXIBLE_PROGRESS_UPDATE byte:", state.bytesDownloaded().toString());
        Log.d("FLEXIBLE_PROGRESS_UPDATE total byte:", state.totalBytesToDownload().toString());

        val byteData = InGooglePlayByteData(
          state.bytesDownloaded().toString(),
          state.totalBytesToDownload().toString()
        );

        val byteJsonString = Gson().toJson(byteData)

        appUpdateEventSink?.success(byteJsonString);

        updateResult?.success(null)
        updateResult = null
      } else if (state.installErrorCode() != InstallErrorCode.NO_ERROR) {
        updateResult?.error(
          "Error during installation",
          state.installErrorCode().toString(),
          null
        )
        updateResult = null
      }
      // Log state or install the update.
    }

    appUpdateManager?.registerListener(installStateUpdatedListener);


//    appUpdateManager?.registerListener { state ->
//      if (state.installStatus() == InstallStatus.DOWNLOADED) {
//        updateResult?.success(null)
//        updateResult = null
//      } else if (state.installErrorCode() != InstallErrorCode.NO_ERROR) {
//        updateResult?.error(
//          "Error during installation",
//          state.installErrorCode().toString(),
//          null
//        )
//        updateResult = null
//      }
//    }
  }

  private fun completeFlexibleUpdate(result: Result) = checkAppState(result) {
    appUpdateManager?.completeUpdate()
  }

  private fun checkForUpdate(result: Result) {
    requireNotNull(activityProvider?.activity()) {
      result.error("REQUIRE_FOREGROUND_ACTIVITY", "in_googleplay_update requires a foreground activity", null)
    }

    activityProvider?.addActivityResultListener(this)
    activityProvider?.activity()?.application?.registerActivityLifecycleCallbacks(this)

    appUpdateManager = AppUpdateManagerFactory.create(activityProvider!!.activity()!!)

    // Returns an intent object that you use to check for an update.
    val appUpdateInfoTask = appUpdateManager!!.appUpdateInfo

    // Checks that the platform will allow the specified type of update.
    appUpdateInfoTask.addOnSuccessListener { info ->
      appUpdateInfo = info
      result.success(
        mapOf(
          "updateAvailability" to info.updateAvailability(),
          "immediateAllowed" to info.isUpdateTypeAllowed(AppUpdateType.IMMEDIATE),
          "flexibleAllowed" to info.isUpdateTypeAllowed(AppUpdateType.FLEXIBLE),
          "availableVersionCode" to info.availableVersionCode(), //Nullable according to docs
          "installStatus" to info.installStatus(),
          "packageName" to info.packageName(),
          "clientVersionStalenessDays" to info.clientVersionStalenessDays(), //Nullable according to docs
          "updatePriority" to info.updatePriority()
        )
      )
    }
    appUpdateInfoTask.addOnFailureListener {
      result.error("TASK_FAILURE", it.message, null)
    }
  }
}


example/lib/main.dart
import 'dart:async';
import 'dart:convert';
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:in_googleplay_update/in_googleplay_update.dart';
import 'package:package_info_plus/package_info_plus.dart';
import 'package:filesize/filesize.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  PackageInfo _packageInfo = PackageInfo(
    appName: 'Unknown',
    packageName: 'Unknown',
    version: 'Unknown',
    buildNumber: 'Unknown',
    buildSignature: 'Unknown',
    installerStore: 'Unknown',
  );
  @override
  void initState() {
    super.initState();
    _initPackageInfo();
  }
  Future<void> _initPackageInfo() async {
    final info = await PackageInfo.fromPlatform();
    setState(() {
      _packageInfo = info;
    });
  }
  Widget _infoTile(String title, String subtitle) {
    return ListTile(
      title: Text(title),
      subtitle: Text(subtitle.isEmpty ? 'Not set' : subtitle),
    );
  }
  AppPlayUpdateInfo? _updateInfo;
  AppUpdateResult? _updateResult;

  GlobalKey<ScaffoldState> _scaffoldKey = new GlobalKey();

  bool _flexibleUpdateAvailable = false;

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> checkForUpdate() async {
    InAppPlayUpdate.checkForUpdate().then((info) {
      setState(() {
        _updateInfo = info;
      });
    }).catchError((e) {
      showSnack(e.toString());
    });
  }

  void showSnack(String text) {
    if (_scaffoldKey.currentContext != null) {
      ScaffoldMessenger.of(_scaffoldKey.currentContext!)
          .showSnackBar(SnackBar(content: Text(text)));
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        key: _scaffoldKey,
        appBar: AppBar(
          title: const Text('In App Update Example App'),
        ),
        body: SingleChildScrollView(
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Column(
              children: <Widget>[
                Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    _infoTile('update availability', '${_updateInfo?.updateAvailability}'),
                    _infoTile('immediate update allowed', '${_updateInfo?.immediateUpdateAllowed}'),
                    _infoTile('available version code', '${_updateInfo?.availableVersionCode}'),
                    _infoTile('install status', '${_updateInfo?.installStatus}'),
                    _infoTile('package name', '${_updateInfo?.packageName}'),
                    _infoTile('client version stanleness days', '${_updateInfo?.clientVersionStalenessDays}'),
                    _infoTile('update priority', '${_updateInfo?.updatePriority}'),
                    _infoTile('update result', '${_updateResult}'),
                  ],
                ),
                ElevatedButton(
                  child: const Text('Check for update'),
                  onPressed: () => checkForUpdate(),
                ),
                ElevatedButton(
                  child: Text('Perform immediate update'),
                  onPressed: _updateInfo?.updateAvailability ==
                      UpdateAvailability.updateAvailable
                      ? () {
                    InAppPlayUpdate.performImmediateUpdate()
                        .catchError((e) => showSnack(e.toString()));
                  }
                      : null,
                ),
                ElevatedButton(
                  child: Text('Start flexible update'),
                  onPressed: _updateInfo?.updateAvailability ==
                      UpdateAvailability.updateAvailable
                      ? () {
                    InAppPlayUpdate.startFlexibleUpdate().then((_) {
                      setState(() {
                        _updateResult = _;
                        _flexibleUpdateAvailable = true;
                      });
                    }).catchError((e) {
                      showSnack(e.toString());
                    });
                  }
                      : null,
                ),
                StreamBuilder<String>(
                  stream: InAppPlayUpdate.flexibleUpdateStream(),
                  builder: (context, snapshot) {
                    if (snapshot.hasData) {
                      String byteStringDownload = snapshot.data ?? '{}';
                      final byteDownload = json.decode(byteStringDownload);
                      return Column(
                        children: [
                          if(int.parse(byteDownload['bytes_downloaded']) == 0 && int.parse(byteDownload['total_bytes_to_download']) == 0)...[
                            ElevatedButton(
                              onPressed: () {
                                InAppPlayUpdate.completeFlexibleUpdate().then((_) {
                                  showSnack("Success!");
                                }).catchError((e) {
                                  showSnack(e.toString());
                                });
                              },
                              child: const Text('Complete flexible update'),
                            ),
                          ] else...[
                            LinearProgressIndicator(
                              value: (int.parse(byteDownload['bytes_downloaded']) / int.parse(byteDownload['total_bytes_to_download'])),
                            ),
                            Text('${filesize(int.parse(byteDownload['bytes_downloaded']))}/${filesize(int.parse(byteDownload['total_bytes_to_download']))}',
                              style: const TextStyle(
                                  fontSize: 14
                              ),
                            )
                          ]
                        ],
                      );
                      // return Text(
                      //   '${filesize(int.parse(byteDownload['bytes_downloaded']))}',
                      //   style: TextStyle(
                      //     fontSize: 10
                      //   ),
                      // );
                    } else {
                      return Column(
                        children: [
                          SizedBox(height: 10),
                          SizedBox(
                            width: 30,
                            height: 30,
                            child: CircularProgressIndicator(),
                          ),
                          Text('This loading for download flexible indicator',
                            style: const TextStyle(
                                fontSize: 14
                            ),
                          )
                        ],
                      );
                    }
                  },
                ),
                Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    _infoTile('App name', _packageInfo.appName),
                    _infoTile('Package name', _packageInfo.packageName),
                    _infoTile('App version', _packageInfo.version),
                    _infoTile('Build number', _packageInfo.buildNumber),
                    _infoTile('Build signature', _packageInfo.buildSignature),
                    _infoTile(
                      'Installer store',
                      _packageInfo.installerStore ?? 'not available',
                    ),
                  ],
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}